contracts/contracts/coordination/BqETHSubscription.sol

```diff
@@ -124,23 +181,64 @@ contract BqETHSubscription is IFeeModel {
+        return baseFeeRate * packageDuration * maxNodes;
+    }
+
+    function encryptorFees(uint128 encryptorSlots, uint32 duration) public view returns (uint256) {
+        return encryptorFeeRate * duration * encryptorSlots;
     }
 
     /**
      * @notice Pays for a subscription
+     * @param encryptorSlots Number of slots for encryptors
      */
-    function payForSubscription() external {
-        // require(endOfSubscription == 0, "Subscription already payed");
-        uint256 amount = packageFees();
-        if (endOfSubscription == 0) {
-            endOfSubscription = uint32(block.timestamp);
+    function payForSubscription(uint128 encryptorSlots) external {
+        uint256 currentPeriodNumber = getCurrentPeriodNumber();
+        require(!billingInfo[currentPeriodNumber + 1].paid, "Next billing period already paid"); // TODO until we will have refunds
+        require(
+            startOfSubscription == 0 ||
+                getEndOfSubscription() + yellowPeriodDuration + redPeriodDuration >=
+                block.timestamp,
+            "Subscription is over"
+        );
+
+        uint256 fees = baseFees() + encryptorFees(encryptorSlots, packageDuration);
+        uint256 periodNumber = currentPeriodNumber;
+        if (billingInfo[periodNumber].paid) {
+            periodNumber++;
         }
-        endOfSubscription += uint32(maxDuration);
+        Billing storage billing = billingInfo[periodNumber];
+        billing.paid = true;
+        billing.encryptorSlots = encryptorSlots;
 
-        feeToken.safeTransferFrom(msg.sender, address(this), amount);
-        emit SubscriptionPaid(msg.sender, amount, endOfSubscription);
+        feeToken.safeTransferFrom(msg.sender, address(this), fees);
+        emit SubscriptionPaid(msg.sender, fees, encryptorSlots, getEndOfSubscription());
+    }
+
+    /**
+     * @notice Pays for additional encryptor slots in the current period
+     * @param additionalEncryptorSlots Additional number of slots for encryptors
+     */
+    function payForEncryptorSlots(uint128 additionalEncryptorSlots) external {
+        uint256 currentPeriodNumber = getCurrentPeriodNumber();
+        Billing storage billing = billingInfo[currentPeriodNumber];
+        require(billing.paid, "Current billing period must be paid");
+
+        uint32 duration = packageDuration;
+        uint32 endOfCurrentPeriod = 0;
+        if (startOfSubscription != 0) {
+            endOfCurrentPeriod = uint32(
+                startOfSubscription + (currentPeriodNumber + 1) * packageDuration
+            );
+            duration = endOfCurrentPeriod - uint32(block.timestamp);
+        }
+
+        uint256 fees = encryptorFees(additionalEncryptorSlots, duration);
+        billing.encryptorSlots += additionalEncryptorSlots;
+
+        feeToken.safeTransferFrom(msg.sender, address(this), fees);
+        emit EncryptorSlotsPaid(msg.sender, fees, additionalEncryptorSlots, endOfCurrentPeriod);
     }
 
     /**
```

Comment:
For the `withdrawToBeneficiary()` method, what do you think if we add the `token` parameter, so that the Treasury can act on any user error when sending a different currency?

Reply:
I'm not against it, but feels like a bit too much. How often it was with other contracts?

Reply:
It hasn't happened yet. Since this is going to be upgradeable, let's pass on this suggestion for the moment.